preserve residual in transformer norm_first

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61692 preserve residual in transformer norm_first**

Fixes #61165, code refactored a bit for clarity.

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

Commit:
preserve residual in transformer norm_first

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

Commit:
Update on "preserve residual in transformer norm_first"


Fixes #61165

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

[ghstack-poisoned]

Commit:
Update on "preserve residual in transformer norm_first"


Fixes #61165

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

[ghstack-poisoned]

Commit:
Update on "preserve residual in transformer norm_first"


Fixes #61165

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

Commit:
Update on "preserve residual in transformer norm_first"


Fixes #61165

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

Commit:
Update on "preserve residual in transformer norm_first"


Fixes #61165

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

[ghstack-poisoned]

cpuhrsch commented on torch/nn/modules/transformer.py:
```diff
@@ -321,25 +321,28 @@ def forward(self, src: Tensor, src_mask: Optional[Tensor] = None, src_key_paddin
+
+        # self-attention block
+        def mha(x):
+            x = self.self_attn(x, x, x, attn_mask=src_mask, key_padding_mask=src_key_padding_mask)[0]
```
nit: need_weights could be set to False if you're indexing at 0 for the return value

bhosmer replied:
Ah yeah nice catch!

Commit:
Update on "preserve residual in transformer norm_first"


Fixes #61165, code refactored a bit for clarity.

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

Commit:
Update on "preserve residual in transformer norm_first"


Fixes #61165, code refactored a bit for clarity.

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

[ghstack-poisoned]

Commit:
Update on "preserve residual in transformer norm_first"


Fixes #61165, code refactored a bit for clarity.

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

jbschlosser commented on torch/nn/modules/transformer.py:
```diff
@@ -321,25 +321,31 @@ def forward(self, src: Tensor, src_mask: Optional[Tensor] = None, src_key_paddin
+        # see Fig. 1 of https://arxiv.org/pdf/2002.04745v1.pdf
+
+        # self-attention block
+        def mha(x):
+            x = self.self_attn(x, x, x,
+                               attn_mask=src_mask,
+                               key_padding_mask=src_key_padding_mask,
+                               need_weights=False)[0]
+            return self.dropout1(x)
+
+        # feed forward block
+        def ff(x):
+            x = self.linear2(self.dropout(self.activation(self.linear1(x))))
+            return self.dropout2(x)
```
Looks like these function definitions break JIT :(

Could try adding them as private functions instead (e.g. `_mha_block()`, `_ffn_block()`)

bhosmer replied:
argh. ok made them methods, that seems more orderly than private functions. But maybe there's a reason why private functions would be better? Happy to change if so, my preference is just aesthetic.

jbschlosser replied:
Ah yeah, sorry I should have said "private methods". I think they're good as methods, but they should be marked as private with the leading underscore so no one depends on them if we want to refactor later on

jbschlosser reviewed (commented):
Thanks for the fix! Does a similar update need to be made in `TransformerDecoderLayer` as well?

Commit:
Update on "preserve residual in transformer norm_first"


Fixes #61165, code refactored a bit for clarity.

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

Commit:
Update on "preserve residual in transformer norm_first"


Fixes #61165, code refactored a bit for clarity.

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

Commit:
Update on "preserve residual in transformer norm_first"


Fixes #61165, code refactored a bit for clarity.

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

jbschlosser reviewed (approved):
LGTM! More correct & much more readable :)

Commit:
Update on "preserve residual in transformer norm_first"


Fixes #61165, code refactored a bit for clarity.

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

navvye commented:
need_weights always should have been set as a parameter for the _sa_block function. This unintentionally made outputting attention_maps using TransformerEncoderLayers impossible.
